File: 143-reorder-list/reorder-list.ts
/**
 * Definition for singly-linked list.
 * class ListNode {
 *     val: number
 *     next: ListNode | null
 *     constructor(val?: number, next?: ListNode | null) {
 *         this.val = (val===undefined ? 0 : val)
 *         this.next = (next===undefined ? null : next)
 *     }
 * }
 */

/**
 Do not return anything, modify head in-place instead.
 */
function reorderList(head: ListNode | null): void {
    // stack push n
    let stack = new Array<ListNode>();
    let node = head;
    while(node){
        stack.push(node);
        node = node.next;
    }
    
    let pre = new ListNode(0);
    // head node.next point to the stack.pop 
    // until node === stack.pop
    while(stack.length != 0){
        
        node = stack.shift();
        console.log("node: " + node.val);
        node.next = stack.pop() ?? null;
        console.log("node.next:" +  node.next?.val);
        pre.next = node;
        pre = node.next;
        console.log("pre:" +  pre?.val);
    }
    if(pre) pre.next = null;
};

File: 143-reorder-list/reorder-list.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */
class Solution {
    public void reorderList(ListNode head) {
        ListNode fast = head;
        ListNode slow = head;
        ListNode node = head;
        Deque<ListNode> stack = new ArrayDeque<>();
        // 1 2 3 4
        // 1 2 3 4 5

        // 1 2 3      4 5   
        // 1 5 2 4 3    4
        // get the second half
        while (fast != null && fast.next != null) {
            slow = slow.next;
            fast = fast.next.next;
        }
        if (fast != null)
            slow = slow.next;
        while (slow != null) {
            stack.push(slow);
            slow = slow.next;
        }

        // insert
        while (node != null && !stack.isEmpty()) {
            ListNode next = node.next;
            ListNode peak = stack.pop();
            node.next = peak;
            peak.next = next;
            node = next;
        }
        if (node != null)
            node.next = null;
    }
}